aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -0,0 +1,258 @@
+#include <ATen/native/Indexing.h>
+#include <ATen/native/IndexingUtils.h>
+
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/AccumulateType.h>
+
+#include <THC/THCDeviceUtils.cuh>
+#include <THC/THCGeneral.h>
+#include <THC/THCTensorSort.cuh>
+#include <ATen/cuda/CUDAContext.h>
+#include <THC/THCThrustAllocator.cuh>
+#include <thrust/execution_policy.h>
+#include <thrust/sort.h>
+
+
+namespace {
+
+#ifdef __HIP_PLATFORM_HCC__
+static const int WARP_SIZE = 64;
+#else
+static const int WARP_SIZE = 32;
+#endif
+
+
+
+
+template <typename scalar_t, int SZ>
+__global__ void indexing_backward_kernel(
+  int64_t* input, int64_t* indices, scalar_t* grad_output, scalar_t* grad_weight,
+  int64_t numel, int64_t stride, int64_t stride_before, int64_t outer_dim) {
+//numel is total number of flattened indices, not expanded to dimensions that are not indexed.
+//stride is the cumulative size of the not-indexed last dimensions
+//stride_before is the stride of the dimension immediately preceding first indexed dimension
+//if indexing starts from the 0th dimension, stride_before does not matter because blockIdx.z will be 0 in thei case
+//outer_dim is number of elements in the first unindexed dimensions
+  using accscalar_t = at::acc_type<scalar_t, true>;
+  int idx = blockIdx.x * blockDim.y + threadIdx.y;
+
+  // Each warp is responsible for an input into the LookupTable.
+  // If the preceding input has the same as this input, then the warp
+  // exits immediately. The warp also processes subsequent inputs with the
+  // same value.
+  //
+  // Input Warp
+  // 1     <warp 1>
+  // 1     <warp 1> (<warp 2> exits without doing any work)
```

Comment:
I guess, the indexes can have the same value when our slice size is greater than 1?